closed/src/jdk.crypto.ec/share/classes/sun/security/ec/NativeXDHKeyAgreement.java

```diff
@@ -0,0 +1,320 @@
+import java.util.function.Function;
+
+import javax.crypto.KeyAgreementSpi;
+import javax.crypto.SecretKey;
+import javax.crypto.ShortBufferException;
+import javax.crypto.spec.SecretKeySpec;
+
+import jdk.crypto.jniprovider.NativeCrypto;
+
+import sun.security.x509.X509Key;
+
+public class NativeXDHKeyAgreement extends KeyAgreementSpi {
+    private static NativeCrypto nativeCrypto;
+    private static final boolean nativeCryptTrace = NativeCrypto.isTraceEnabled();
+
+    private XECPrivateKey xecPrivateKey;
+    private byte[] privateKey;
+    private byte[] secret;
+    private XECOperations ops;
+    private final XECParameters lockedParams;
+
+    private XDHKeyAgreement javaImplementation;
+
+    public NativeXDHKeyAgreement() {
+        lockedParams = null;
+    }
+
+    public NativeXDHKeyAgreement(AlgorithmParameterSpec paramSpec) {
+        lockedParams = XECParameters.get(ProviderException::new, paramSpec);
+    }
+
+    @Override
+    protected void engineInit(Key key, SecureRandom random)
+            throws InvalidKeyException {
+
+        initImpl(key);
+    }
+
+    @Override
+    protected void engineInit(Key key, AlgorithmParameterSpec params,
+                              SecureRandom random) throws InvalidKeyException,
+        InvalidAlgorithmParameterException {
+
+        initImpl(key);
+
+        // the private key parameters must match params, if present
+        if (params != null) {
+            XECParameters xecParams = XECParameters.get(
+                InvalidAlgorithmParameterException::new, params);
+            if (!xecParams.oidEquals(this.ops.getParameters())) {
+                throw new InvalidKeyException("Incorrect private key parameters");
+            }
+        }
+    }
+
+    private void initImpl(Key key) throws InvalidKeyException {
+
+        if (!(key instanceof XECPrivateKey)) {
+            throw new InvalidKeyException("Unsupported key type");
+        }
+        xecPrivateKey = (XECPrivateKey) key;
+        XECParameters xecParams = XECParameters.get(
+            InvalidKeyException::new, xecPrivateKey.getParams());
+
+        if ((lockedParams != null) && (lockedParams != xecParams)) {
+            throw new InvalidKeyException("Parameters must be " + lockedParams.getName());
+        }
+
+        this.ops = new XECOperations(xecParams);
+        this.privateKey = xecPrivateKey.getScalar()
+                .orElseThrow(() -> new InvalidKeyException("No private key value")
+        );
+        secret = null;
+
+        if (!NativeCrypto.isAllowedAndLoaded()) {
+            if (nativeCryptTrace) {
+                System.err.println("OpenSSL library not loaded." +
+                    " Using Java crypto implementation to generate secret.");
+            }
+            initializeJavaImplementation();
+        } else {
+            /* Assign ID used by OpenSSL for different curves. */
+            int curveType;
+            if ("X25519".equalsIgnoreCase(ops.getParameters().getName())) {
```

Comment:
I suggest this should not ignore case, and the test appears often enough to warrant a helper method:
```java
    private static boolean isX25519(XECParameters parameters) {
        return "X25519".equals(parameters.getName());
    }
```